Quadkey module

vdelacruzb opened the pull request:
Wanted to check with you before you go on holidays, at least the basics and name of functions.

- I created some integration tests using for loops but I set the step really high to avoid long waiting periods. The step in js checks is lower to compare more values.

- The doc is very basic. There is not much into these functions just some conversions and they are public functions so the doc is more for users and less for devs. Will have to iterate on it and see with product as they will generate also public documentation.

- I have to iterate in the implementation of parent and children, run out of time. They work fine, but it should be relatively easy to do it without decoding encoding.

- Also have to explore more deeply how the mapbox tile cover works. It converts GeoJSON to Quadkeys, and I guess it should be possible to directly convert to quadint and skip steps.

Algunenano commented on quadkey/README.md:
```diff
@@ -0,0 +1,12 @@
+# Quadkey
+
+This contains all the necessary code to build, generate, test and deploy the [BigQuery User Defined Functions](https://cloud.google.com/bigquery/docs/reference/standard-sql/user-defined-functions) for interacting with quadkey indexes.
```
This shouldn't mention BQ here, since we will have other database support.

Algunenano commented on quadkey/bq/Makefile:
```diff
@@ -0,0 +1,77 @@
+POST_INTEGRATION_CLEANUP ?= 1
+
+# Deployment variables
+BQ_LIBRARY_QUADKEY ?= $(BQ_BUCKET_PUBLIC)$(BQ_DATASET_QUADKEY)/quadkey_library.js
```
SKEL library is named `SKEL_BQ_LIBRARY` so this should be `QUADKEY_BQ_LIBRARY`

Algunenano commented on quadkey/bq/sql/BBOX.sql:
```diff
@@ -0,0 +1,16 @@
+-----------------------------------------------------------------------
+--
+-- Copyright (C) 2020 - 2021 CARTO
```
Copyrights should be 2021 only, right?

Algunenano reviewed (commented):
Some notes from a quick preview (not in depth):
* There shouldn't be parseInt / Number calls in the JS functions just because BigQuery require them, that should be part of the BQ API, not the library itself.
* I see a LICENSE but i'm worried it's added ad-hoc and not the real license.
* Related to the previous point, where does the code from `quadkey.js` comes from. What part comes from other project and how do we update it? What part is added by us? How do we mix the 2 together?
* The style of the JS tests is completely different from the code. Maybe we should lint the tests too? And maybe a common .eslint file somewhere in the repo :thinking: ...

vdelacruzb commented:
> * I see a LICENSE but i'm worried it's added ad-hoc and not the real license.
> * Related to the previous point, where does the code from `quadkey.js` comes from. What part comes from other project and how do we update it? What part is added by us? How do we mix the 2 together?

Regarding these two topics, I basically took the same code that we had in jslibs -> quadkey.js. And then adapted it to support quadints instead of quadkeys by updating the functions that used quadkeys and also by adding parser from quadkey to quadints. So it should be under the same terms of license than the past code. Was the last code taken from anywhere or totally created here? 

I would say that the quadint part doesn't have other sources. I calculated that we needed 5 bits for zoom and then when saw this https://github.com/joekarl/binary-quadkey it was a verification, but we didn't follow that github implementation.

Inside tilecover.js we use mapbox/tilecover. Does it affect our license?

Algunenano commented on quadkey/bq/test/quadkey_integration.js:
```diff
@@ -0,0 +1,171 @@
+
+    it ('Should be able to encode/decode quadints at different zooms', async () => {
+        let tilesPerLevel, x, y;
+        for(let z = 1; z < 30; z = z + 7)
```
Please don't do a single query per zoom level, this is going to be awfully slow. Instead compose a query with all the things you want to check and send a single query to BigQuery and then check the results. You should also add tiles in between the first and last, and check what happens when invalid values are passed (if we want to handle it).

Same for ALL other integration tests.

Algunenano commented on quadkey/library/test/quadkey_test.js:
```diff
@@ -0,0 +1,133 @@
+                }
+            }
+        }
+    });
```
Missing sibling tests.

Algunenano commented on quadkey/bq/test/quadkey_integration.js:
```diff
@@ -0,0 +1,171 @@
+                    assert.equal(cont,1);
+                }
+            }
+        }
```
Where are the tests for all the other functions (`ST_ASQUADINTPOLYFILL`, `ST_GEOGFROMQUADINT_BOUNDARY`, `KRING`...)? Is there something missing in this PR?

Algunenano commented on quadkey/library/js/tilecover.js:
```diff
@@ -0,0 +1,585 @@
+(function() {
```
Where is the copyright, license and notes about where this comes from? We don't have to add other stuff to files that come from other repositories if we copy them as-is, but we should be extremely clear about it in the license notices and document where they come from, what they do and how to update them.

Algunenano reviewed (changes requested):
> Regarding these two topics, I basically took the same code that we had in jslibs -> quadkey.js. And then adapted it to support quadints instead of quadkeys by updating the functions that used quadkeys and also by adding a parser from quadkey to quadints and viceversa. So it should be under the same terms of license than the past code. Was the last code taken from anywhere or totally created here?

This is exactly what needs to be figured out as part of this PR. Find out where the older code comes from, whether we can use it or not, document it and explain how to update and so on. You can't just add whatever license we have for the general project to external code.

vdelacruzb commented:
Did some refactoring and avoided using totally the old code due to possible license uses.
Also removed tilecover.js as it was a browserified version of @mapbox/tilebelt and @mapbox/tile-cover with some of our code. Now the makefile directly browserify the quadint code with these two modules which already contain auxiliar functions to make it work.
I added the LICENSE of the modules to the js folder.

Algunenano commented on quadkey/bq/Makefile:
```diff
@@ -0,0 +1,77 @@
+		-e 's!@@QUADKEY_BQ_LIBRARY@@!$(QUADKEY_BQ_LIBRARY)!g'
+
+dataset_deploy: check_environment
+	for n in $(wildcard sql/*.sql); do \
```
```suggestion
	for n in $(sort $(wildcard sql/*.sql)); do \
```

Algunenano commented on quadkey/bq/sql/ST_ASQUADINTPOLYFILL.sql:
```diff
@@ -0,0 +1,28 @@
+--
+-----------------------------------------------------------------------
+
+CREATE OR REPLACE FUNCTION `@@BQ_PROJECTID@@.@@BQ_DATASET_QUADKEY@@.POLYFILL_FROMGEOJSON`
```
I understand that this function is private right? In that case I would name it as __ST_ASQUADINT_POLYFILL

Algunenano commented on quadkey/bq/sql/ST_ASQUADINTPOLYFILL.sql:
```diff
@@ -0,0 +1,28 @@
+    return quadints.map(String);
+""";
+
+CREATE OR REPLACE FUNCTION `@@BQ_PROJECTID@@.@@BQ_DATASET_QUADKEY@@.ST_ASQUADINTPOLYFILL`
```
```suggestion
CREATE OR REPLACE FUNCTION `@@BQ_PROJECTID@@.@@BQ_DATASET_QUADKEY@@.ST_ASQUADINT_POLYFILL`
```

For consistency with `ST_ASH3_POLYFILL`

Algunenano commented on quadkey/bq/sql/ST_GEOGFROMQUADINT_BOUNDARY.sql:
```diff
@@ -0,0 +1,27 @@
+--
+-----------------------------------------------------------------------
+
+CREATE OR REPLACE FUNCTION `@@BQ_PROJECTID@@.@@BQ_DATASET_QUADKEY@@.GEOJSONBOUNDARY_FROMQUADINT`
```
Same, this is private, so I would call it as the main function and add the `__` prefix.

Algunenano commented on quadkey/bq/doc/REFERENCE.md:
```diff
@@ -0,0 +1,148 @@
+
+Transform a quadint index to an equivalent quadkey.
+
+#### quadkey.PARENT
```
In H3 this is called `TOPARENT` and supports passing the desired resolution. Thoughts?

vdelacruzb replied:
This should be doable on my side. I go for it.

Algunenano commented on quadkey/bq/doc/REFERENCE.md:
```diff
@@ -0,0 +1,148 @@
+
+Returns the parent quadint of a given quadint. A parent quadint is a quadint of smaller level of detail (zoom - 1) which contains the current quadint.
+
+#### quadkey.CHILDREN
```
Same, in `h3` this function is called `TOCHILDREN` and supports passing the resolution of the children you want.

vdelacruzb replied:
This should be also doable. I guess this should return all the children contained for a given resolution right? Example, current quadint is in resolution 1, TOCHILDREN(quadint, 3)-> array with 16 children.

Algunenano commented on quadkey/library/js/quadint.js:
```diff
@@ -0,0 +1,190 @@
+    }
+    if (direction === 'down') {
+        y = y < tilesPerLevel - 1 ? y + 1 : 0;
+    }
```
I think checking the direction is one of the valid values here would be best, otherwise you get the same input and no error (e.g. "Left" won't work as expected).

vdelacruzb replied:
oki doki. I'm going to check for expected values. Do you like the idea of doing also a toLowerCase? or should be less flexible and not allow capital letters?

Algunenano replied:
Doing `toLowerCase` seems llike a good idea

Algunenano commented on quadkey/library/test/quadkey_test.js:
```diff
@@ -0,0 +1,140 @@
+        }
+    });
+
+    it('Should be able to encode/decode between quadint and quadkey at any level of zoom', async() => {
```
I think I would add some known value tests and not only rountrip, otherwise you risk breaking compatibility and not noticing it.

Algunenano commented on quadkey/bq/test/quadkey_integration.js:
```diff
@@ -0,0 +1,500 @@
+const BQ_PROJECTID = process.env.BQ_PROJECTID;
+const BQ_DATASET_QUADKEY = process.env.BQ_DATASET_QUADKEY;
+
+describe('QUADKEY integration tests', () => {
```
Personally I prefer to break down the integration tests in multiple files, it's clearer to see what's being tested in each of them and they are executed in parallel (so everything is much faster). In any case, this looks good.

vdelacruzb replied:
I will split it then as it should be a fast change.

Algunenano reviewed (commented):
Looks great. I've added some comments here and there about possible improvements, mostly to keep things consistent now that we have more modules added (so it could be changed in the other modules instead).

Algunenano reviewed (approved):
Looks good
